stores/metadata.go

```diff
@@ -1482,33 +1482,51 @@ func (s *SQLStore) UpdateObject(ctx context.Context, bucket, path, contractSet,
 }
 
 func (s *SQLStore) RemoveObject(ctx context.Context, bucket, path string) error {
-	var rowsAffected int64
-	var err error
-	err = s.retryTransaction(ctx, func(tx *gorm.DB) error {
-		rowsAffected, err = s.deleteObject(tx, bucket, path)
-		if err != nil {
-			return fmt.Errorf("RemoveObject: failed to delete object: %w", err)
-		}
-		return nil
+	var prune bool
+	err := s.bMain.Transaction(ctx, func(tx sql.DatabaseTx) (err error) {
+		prune, err = tx.DeleteObject(ctx, bucket, path)
+		return
 	})
 	if err != nil {
-		return err
-	}
-	if rowsAffected == 0 {
+		return fmt.Errorf("RemoveObject: failed to delete object: %w", err)
+	} else if !prune {
 		return fmt.Errorf("%w: key: %s", api.ErrObjectNotFound, path)
 	}
+	s.triggerSlabPruning()
 	return nil
 }
 
 func (s *SQLStore) RemoveObjects(ctx context.Context, bucket, prefix string) error {
-	var err error
-	deleted, err := s.deleteObjects(ctx, bucket, prefix)
-	if err != nil {
-		return err
+	var prune bool
+	batchSizeIdx := 0
+	for {
+		start := time.Now()
+		done := false
+		var duration time.Duration
+		if err := s.bMain.Transaction(ctx, func(tx sql.DatabaseTx) error {
+			deleted, err := tx.DeleteObjects(ctx, bucket, prefix, objectDeleteBatchSizes[batchSizeIdx])
+			if err != nil {
+				return err
+			}
+			prune = prune || deleted
+			done = !deleted
+			return nil
+		}); err != nil {
+			return fmt.Errorf("failed to delete objects: %w", err)
+		} else if done {
+			break // nothing more to delete
+		}
+		duration = time.Since(start)
+
+		// increase the batch size if deletion was faster than the threshold
+		if duration < batchDurationThreshold && batchSizeIdx < len(objectDeleteBatchSizes)-1 {
```

Comment:
I know it's probably premature to optimise this but we could also decrement when we're over the threshold and stay there for the rest of the batch. I would usually never do that but I'm wondering whether it makes sense here for UI/UX reasons. E.g. batching all removals in batches of 10 might really be twice as fast as trying to do batches of 50.

Reply:
Tbh I want to dedicate a whole PR to this including a benchmark. I'm not sure if we even need batching anymore here. Because we used to do object deletion and slab pruning in the same txn and the latter was the slow operation. Now that we only delete objects it might be fine to just not batch.